feat(values): Add flag `--values-directory`

Bhargav-InfraCloud opened the pull request:
**What this PR does / why we need it**:

Introduced new CLI flag `-d`/`--values-directory` for loading values from YAML files from a specified directory. Following are the key details:

- This will be used by the following commands:
  1. `install`
  2. `lint`
  3. `package`
  4. `template`
  5. `upgrade`
  Updated the help text of these commands with the same.

- Made this flag the lowest in precedence i.e., values from values directory can be overridden by inputs from any other input flags (`-f`/`--values`, `--set-json`, `--set`, `--set-string`, `--set-file`, `--set-literal`).
  > Note: The default values from the chart's `values.yaml` file will have the lowest precedence than all the input flags including `-d`/`--values-directory` by design. In order the values from `values.yaml` to override the values from `-d`/`--values-directory`, users can explicitly provide the chart's `values.yaml` file path using the `-f`/`--values` flag.

- This flag reads all `.yaml` files from the specified directory and its subdirectories recursively, in lexicographical order. For eg. for the following directory structure:
    ```
    foo/
    ├── bar/
    │    └── bar.yaml
    ├── baz/
    │    ├── baz.yaml
    │    └── qux.yaml
    ├── baz.txt
    └── foo.yaml
    ```
  The files will be read in the order: `bar/bar.yaml`, `baz/baz.yaml`, `baz/qux.yaml`, `foo.yaml`.

- If the specified directory does not exist, an error is returned.

- Non-YAML files in the directory are ignored. i.e., if the specified directory exists but contains no YAML files, an empty map is returned without errors.

- If multiple YAML files contain overlapping keys, the values are overridden in lexicographical order. For eg. in the above example, if both `baz/baz.yaml` and `foo.yaml` contain a key `replicaCount`, the value from `foo.yaml` will take precedence as it is read last.

Fixes #10416

**Special notes for your reviewer**:

**If applicable**:
- [x] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

sirianni commented:
Could we also consider the `values.d` capability for defining default values in the chart itself?  This would be very useful as the default `values.yaml` file in some charts can get quite large and unwieldy.
```
my-chart/
  Chart.yaml
  templates/
  values.d/
    foo.yaml
    bar.yaml
```

Bhargav-InfraCloud commented:
@sirianni That makes sense. Can you go thru #10416 where we are discussing the approaches on this? You can help by giving your thoughts on it.

SytemicAnomaly commented:
Hi @sabre1041, this is the PR we talked about at KubeCon today.
Thanks in advance

Syst3m1cAn0maly commented:
Hi @Bhargav-InfraCloud,
I added a PR to your branch to support directories directly with the --values flag instead of the created --values-dir as I suggested in the Issue [10416](https://github.com/helm/helm/issues/10416)

Syst3m1cAn0maly commented:
Should I create a new PR directly with the helm repo with the modifications I added in the PR to @Bhargav-InfraCloud's repo ?

Bhargav-InfraCloud commented:
Hey @Syst3m1cAn0maly, my apologies. I couldn't reply any sooner. I agree, adding this new behavior to the existing flag `--values` wouldn't break anything. But it is still a changed behavior of one of the most used flags. I think keeping the functionalities separated isn't a bad idea either (at least for now).

We had this as one of the approaches previously discussed in [comment][1].

It's your call if you want to create a separate PR for helm directly. But I had to suggest, my suggestion to you would be to do it for helm v4 instead, regardless of the merge status of this current PR. More details about helm 4 are [here][2].

[1]: https://github.com/helm/helm/issues/10416#issuecomment-1416246208
[2]: https://github.com/helm/helm/blob/main/CONTRIBUTING.md#milestones

Syst3m1cAn0maly commented:
No problem @Bhargav-InfraCloud ;)

Thank you for the explanations, I think I will create a PR for helm 4 as suggested unless there is a change of heart on the maintainers'side

Bhargav-InfraCloud commented:
FYI: Refactored my old changes and did a clean-up of tests.
Please review the PR.

ryan-dyer-sp commented:
@Bhargav-InfraCloud Not really qualified to review, but do have some usability questions.  It appears that the intent is to have separate values files which dont contradict each other within the directory.  but take something like [grafana/loki](https://github.com/grafana/loki/tree/main/production/helm/loki) where they have different values files(distributed & simple scalable) they may publish depending on the scale at which the admin wishes to operate.  How do you see this solution allowing that to happen?  Each of these values files would go into a different subdirectory?  
What about if multiple files within the directory do happen to use the same object defined?  I see that this should be predictable at least in that its using Walk which follows lexical order, but this could lead to confusion if people arent aware of this.
My use case would be similar to the grafana one where we want to be able to include various values files within the helm chart from a packaging perspective and then be able to mix and match them as we see fit.  It would appear that in order to achieve that behavior with this, each file would need to go into a distinct subdir with its own file and then just reference the subdirs via -d like we would via -f if they were locally available.

Bhargav-InfraCloud commented:
Hi @ryan-dyer-sp,
The actual use case: if the values file grows too big to manage, split it into multiple files and group them in a directory, so that managing becomes easier. This is similar to how Linux groups configs (eg. `/etc/sysctl.d`).
Hence the `--values-directory` for the helm to recursively read all the `.yaml` files in the specified directory. 

> My use case would be similar to the grafana one where we want to be able to include various values files within the helm chart from a packaging perspective and then be able to mix and match them as we see fit. It would appear that in order to achieve that behavior with this, each file would need to go into a distinct subdir with its own file and then just reference the subdirs via -d like we would via -f if they were locally available.

Yes, you can specify each directory or file "in the order" so that the latter can replace the former. Multiple sub-directories should also work as long as they are organized in the lexical order.

Note: Values from all other values setting flags (`--values`, `--set-json`, `--set`, `--set-string`, `--set-file`, `--set-literal`) can override the ones from `--values-directory` if matching.

> What about if multiple files within the directory do happen to use the same object defined? I see that this should be predictable at least in that its using Walk which follows lexical order, but this could lead to confusion if people arent aware of this.

Yes, it follows the lexical order. I think it would've been more confusing if it hadn't followed the lexical order 😄 But I agree, it is better to add that detail. Updated the flag description. Thanks for pointing it out.

Bhargav-InfraCloud commented on pkg/cli/values/options.go:
```diff
@@ -145,3 +173,31 @@ func readFile(filePath string, p getter.Providers) ([]byte, error) {
+	var files []string
+
+	// Traverse through the directory, recursively
+	err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
```
Didn't realize there was a better version of it 😅 Updated it now. Thanks!

Bhargav-InfraCloud committed:
feat(values): Add flag -d/--values-directory

Introduced new CLI flag `-d`/`--values-directory`
for loading values from YAML files from a
specified directory. Following are the key
details:

- This will be used by the following commands:
  1. `install`
  2. `lint`
  3. `package`
  4. `template`
  5. `upgrade`
  Updated the help text of these commands with the
  same.

- Made this flag the lowest in precedence i.e.,
  values from values directory can be overridden
  by inputs from any other input flags
  (`-f`/`--values`, `--set-json`, `--set`,
  `--set-string`, `--set-file`, `--set-literal`).
  Note: The default values from the chart's
  `values.yaml` file will have the lowest
  precedence than all the input flags including
  `-d`/`--values-directory` by design. In order
  the values from `values.yaml` to override the
  values from `-d`/`--values-directory`, users can
  explicitly provide the chart's `values.yaml`
  file path using the `-f`/`--values` flag.

- This flag reads all `.yaml` files from the
  specified directory and its subdirectories
  recursively, in lexicographical order. For eg.
  for the following directory structure:

  foo/
  ├── bar/
  │    └── bar.yaml
  ├── baz/
  │    ├── baz.yaml
  │    └── qux.yaml
  ├── baz.txt
  └── foo.yaml

  The files will be read in the order:
  `bar/bar.yaml`, `baz/baz.yaml`, `baz/qux.yaml`,
  `foo.yaml`.

- If the specified directory does not exist, an
  error is returned.

- Non-YAML files in the directory are ignored.
  i.e., if the specified directory exists but
  contains no YAML files, an empty map is returned
  without errors.

- If multiple YAML files contain overlapping keys,
  the values are overridden in lexicographical
  order. For eg. in the above example, if both
  `baz/baz.yaml` and `foo.yaml` contain a key
  `replicaCount`, the value from `foo.yaml` will
  take precedence as it is read last.

Fixes helm#10416

Signed-off-by: Bhargav Ravuri <bhargav.ravuri@infracloud.io>

Bhargav-InfraCloud commented:
> Needs rebase. Note Helm 3 is not accepting new features.

@scottrigby / all,
Rebased the PR on the latest `main` (i.e., v4). Added some more test cases. Please review.

Bhargav-InfraCloud commented on pkg/cli/values/testdata/one-piece-chart/values.d/values-directory-with-single-yaml-file/alliances-law.yaml:
```diff
@@ -0,0 +1 @@
+../shared-values/alliances-law.yaml
```
To prevent duplicates, I used symbolic links in the test data.
